app/src/main/java/org/thoughtcrime/securesms/jobs/BoostReceiptRequestResponseJob.java

```diff
@@ -0,0 +1,230 @@
+                                                                                 .blockingGet();
+
+    if (response.getApplicationError().isPresent()) {
+      handleApplicationError(response);
+    } else if (response.getResult().isPresent()) {
+      ReceiptCredential receiptCredential = getReceiptCredential(response.getResult().get());
+
+      if (!isCredentialValid(receiptCredential)) {
+        throw new IOException("Could not validate receipt credential");
+      }
+
+      ReceiptCredentialPresentation receiptCredentialPresentation = getReceiptCredentialPresentation(receiptCredential);
+      setOutputData(new Data.Builder().putBlobAsString(DonationReceiptRedemptionJob.INPUT_RECEIPT_CREDENTIAL_PRESENTATION,
+                                                       receiptCredentialPresentation.serialize())
+                                      .build());
+    } else {
+      Log.w(TAG, "Encountered a retryable exception: " + response.getStatus(), response.getExecutionError().orNull(), true);
+      throw new RetryableException();
+    }
+  }
+
+  private static void handleApplicationError(ServiceResponse<ReceiptCredentialResponse> response) throws Exception {
+    Throwable applicationException = response.getApplicationError().get();
+    switch (response.getStatus()) {
+      case 204:
+        Log.w(TAG, "User does not have receipts available to exchange. Exiting.", applicationException, true);
+        break;
+      case 400:
+        Log.w(TAG, "Receipt credential request failed to validate.", applicationException, true);
+        throw new Exception(applicationException);
+      case 409:
+        Log.w(TAG, "Receipt already redeemed with a different request credential.", response.getApplicationError().get(), true);
+        throw new Exception(applicationException);
+      default:
+        Log.w(TAG, "Encountered a server failure: " + response.getStatus(), applicationException, true);
+        throw new RetryableException();
+    }
+  }
+
+  private ReceiptCredentialPresentation getReceiptCredentialPresentation(@NonNull ReceiptCredential receiptCredential) throws RetryableException {
+    ClientZkReceiptOperations operations = ApplicationDependencies.getClientZkReceiptOperations();
+
+    try {
+      return operations.createReceiptCredentialPresentation(receiptCredential);
+    } catch (VerificationFailedException e) {
+      Log.w(TAG, "getReceiptCredentialPresentation: encountered a verification failure in zk", e, true);
+      requestContext = null;
+      throw new RetryableException();
+    }
+  }
+
+  private ReceiptCredential getReceiptCredential(@NonNull ReceiptCredentialResponse response) throws RetryableException {
+    ClientZkReceiptOperations operations = ApplicationDependencies.getClientZkReceiptOperations();
+
+    try {
+      return operations.receiveReceiptCredential(requestContext, response);
+    } catch (VerificationFailedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)